include/tvm/runtime/container.h

```diff
@@ -0,0 +1,220 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file tvm/runtime/container.h
+ * \brief Common POD(plain old data) container types.
+ */
+#ifndef TVM_RUNTIME_CONTAINER_H_
+#define TVM_RUNTIME_CONTAINER_H_
+#include <tvm/runtime/object.h>
+#include <initializer_list>
+#include <type_traits>
+#include <vector>
+
+namespace tvm {
+namespace runtime {
+
+/**
+ * @brief Base template for classes with array like memory layout.
+ *
+ *        It provides general methods to access the memory. The memory
+ *        layout is ArrayType + [ElemType]. The alignment of ArrayType
+ *        and ElemType is handled by the memory allocator.
+ *
+ * @tparam ArrayType
+ * @tparam ElemType
+ */
+template <typename ArrayType, typename ElemType>
+class InplaceArrayBase {
+ public:
+  /**
+   * @brief Initialize the elements in the array.
+   */
+  void Init() {
```

Comment:
Think a bit more, what functions do you need to support a dynamic shaped array? where you might need to insert elements  :) ?

Reply:
Oh, I was not aware of this use case. I think it would be great if we can just provide `std::vector` interface(`operator[]`, `push_back`, `size`) in `InplaceArrayBase`. 

To support dynamic insertion, it can query child class with `size()` and `capacity()` in `push_back` implementation. And I won't support dynamically grow the underlying storage since the actual memory management is done by child class, it's also difficult to grow this memory.

Reply:
No we won't support dynamic growth of the object, but perhaps will need to support push_back when capacity() and size() are different. 

Ideally, InplaceArrayBase will be a base class for all containers(Array, ADT String) :)